File: external/grpc/scala/src/main/scala/org/aiddl/external/grpc/scala/sensor/SensorClient.scala
package org.aiddl.external.grpc.scala.sensor

import io.grpc.ManagedChannelBuilder
import org.aiddl.common.scala.execution.sensor.Sensor
import org.aiddl.common.scala.execution.sensor.SensorMode
import org.aiddl.core.scala.container.Container
import org.aiddl.core.scala.parser.Parser
import org.aiddl.core.scala.representation.Term
import org.aiddl.external.grpc.sensor.SensorGrpc
import org.aiddl.external.grpc.empty.Empty
import org.aiddl.external.grpc.scala.converter.Converter

class SensorClient (mode: SensorMode, host: String, port: Int, container: Container) extends Sensor {
  val parser = new Parser(container)
  val converter = new Converter(container)

  override val sensorMode: SensorMode = mode

  val channel = ManagedChannelBuilder
    .forAddress(host, port)
    .usePlaintext()
    .build

  val blockingStub = SensorGrpc.blockingStub(channel)
  override def performSense: Term = {
    val latest = blockingStub.sense(Empty())
    converter.pb2aiddl(latest.getValue)
  }
}

File: external/grpc/python/aiddl_external_grpc_python/sensor/sensor_client.py
import grpc
import aiddl_external_grpc_python.generated.empty_pb2 as empty_pb2

import aiddl_external_grpc_python.generated.sensor_pb2_grpc as sensor_pb2_grpc
from aiddl_external_grpc_python.converter import Converter


class SensorClient:
    def __init__(self, host, port, container):
        channel = grpc.insecure_channel(f'{host}:{port}')
        self.stub = sensor_pb2_grpc.SensorStub(channel)
        self.converter = Converter(container)

    def sense(self):
        answer = self.stub.Sense(empty_pb2.Empty())
        return (self.converter.pb2aiddl(answer.value), answer.sequence_id, answer.timestamp_nano)

    def get_latest_sensor_value(self):
        answer = self.stub.GetLatestSensorValue(empty_pb2.Empty())
        return (self.converter.pb2aiddl(answer.value), answer.sequence_id, answer.timestamp_nano)

